judge/views/api/api_v2.py

```diff
@@ -1,122 +1,593 @@
+                })
+
+        for key, filter_name in self.list_filters:
+            if key in self.request.GET:
+                # May raise ValueError or ValidationError, but is excepted in APIMixin
+                queryset = queryset.filter(**{
+                    filter_name + '__in': self.request.GET.getlist(key),
+                })
+
+        return queryset.distinct()
+
+    def get_queryset(self):
+        return self.filter_queryset(self.get_unfiltered_queryset())
+
+    def get_api_data(self, context):
+        page = context['page_obj']
+        objects = context['object_list']
+        return {
+            'current_object_count': len(objects),
+            'objects_per_page': page.paginator.per_page,
+            'total_objects': page.paginator.count,
+            'page_index': page.number,
+            'total_pages': page.paginator.num_pages,
+            'objects': [self.get_object_data(obj) for obj in objects],
+        }
+
+
+class APIDetailView(APIMixin, BaseDetailView):
+    def get_api_data(self, context):
+        return {
+            'object': self.get_object_data(context['object']),
+        }
+
+
+class APIContestList(APIListView):
+    model = Contest
+    list_filters = (
+        ('tag', 'tags__name'),
+        ('organization', 'organizations'),
+    )
+
+    def get_unfiltered_queryset(self):
+        return (
+            Contest.get_visible_contests(self.request.user)
+            .prefetch_related(
+                Prefetch(
+                    'tags',
+                    queryset=ContestTag.objects.only('name'),
+                    to_attr='tag_list',
+                ),
+            )
+            .order_by('end_time')
+        )
+
+    def get_object_data(self, contest):
+        return {
+            'key': contest.key,
+            'name': contest.name,
+            'start_time': contest.start_time.isoformat(),
+            'end_time': contest.end_time.isoformat(),
+            'time_limit': contest.time_limit and contest.time_limit.total_seconds(),
+            'tags': list(map(attrgetter('name'), contest.tag_list)),
+        }
+
+
+class APIContestDetail(APIDetailView):
+    model = Contest
+    slug_field = 'key'
+    slug_url_kwarg = 'contest'
+
+    def get_object(self, queryset=None):
+        contest = super().get_object(queryset)
+        if not contest.is_accessible_by(self.request.user):
+            raise Http404()
+        return contest
+
+    def get_object_data(self, contest):
+        in_contest = contest.is_in_contest(self.request.user)
+        can_see_rankings = contest.can_see_scoreboard(self.request.user)
+        if contest.hide_scoreboard and in_contest:
+            can_see_rankings = False
+        can_see_problems = (in_contest or contest.ended or contest.is_editable_by(self.request.user))
+
+        problems = list(
+            contest.contest_problems
+            .select_related('problem')
+            .defer('problem__description')
+            .order_by('order'),
```

Comment:
Is the trailing comma necessary?

Reply:
It seems it is necessary for flake to pass.